### ISSUE TITLE
Potential issue with timeout

### ISSUE THREAD

**Issue by gterzian** [servo/webxr#34]
In the `recv_timeout` at 

https://github.com/servo/webxr/blob/35559ae62887648dda0120388462eed4f90e8553/webxr-api/session.rs#L226

Since it's called as part of

**Comment by gterzian** [comment]
Ah no it's not true, since cancelling the raf callback still results in sending a `SessionMsg::RenderAnimationFrame` for each raf, which increments the timestamp and would break the main-thread out of the while loop. 

So if script were to send a raf every 5 ms, and immediately cancel the callback, you'd still break out of the loop when the raf callback, cancelled or not, is executed and a message is sent back to the main-thread. 

However, that requires the callback to execute, which means script has to handle a message. 

So if script were to just start looping in a task, and send a raf every 5 ms, it could keep the main-thead looping and prevent the raf callback from executing by remaining in a long-running task. 

However it would be obvious a script was long-running, so the UA could actually stop it. So while the script is running a long-running task that would send rad every 5 ms, Servo would remain unresponsive to user input, since the main-thread would be stuck in a loop handling raf requests.

**Comment by asajeffrey** [comment]
In order to perform this attack, script would have to create an rAF callback that never ended (in order to block the `RenderAnimationFrame` message) but in that case script wouldn't send another `RequestAnimationFrame` message, and the timeout would expire.

That said, the code is a bit brittle, it might be better to use `recv_deadline` rather than `recv_timeout`.

**Comment by gterzian** [comment]
> In order to perform this attack, script would have to create an rAF callback that never ended, but in that case script wouldn't send another RequestAnimationFrame message, and the timeout would expire.

The way I read `RequestAnimationFrame`

https://github.com/servo/servo/blob/9043f247d9b031ed285e880e4b90aa523d4a63ae/components/script/dom/xrsession.rs#L287

is that it doesn't prevent the currently running task in script from making an arbitrary number of calls, and each call sends a `SessionMsg`, via `Session.request_animation_frame`:

https://github.com/servo/webxr/blob/fcbdddb66e2d307d86ff109d79001c7a45de79db/webxr-api/session.rs#L107

So the attack doesn't so much require a never ending raf callback, it requires a long-running task making calls to `RequestAnimationFrame` every 5 ms. 

Each call sends a message to the main-thread, which resets the timeout, but doesn't increment the timestamp(that requires the raf callback to execute). 

The long-running task prevents raf callback from executing, since they require `xr_raf_callback` task enqueued by the router to executed on the loop, which isn't going to happen so long as another task is keeping the loop busy. 

https://github.com/servo/servo/blob/9043f247d9b031ed285e880e4b90aa523d4a63ae/components/script/dom/xrsession.rs#L310

This could be solved on the script side by requiring a `RequestAnimationFrame` to only queue a callback if the last callback was executed, unfortunately that doesn't appear spec compliant...
https://immersive-web.github.io/webxr/#dom-xrsession-requestanimationframe

**Comment by asajeffrey** [comment]
We might also want to replace `recv_timeout` by `recv_deadline` as a belt-and-braces solution?

**Comment by gterzian** [comment]
> We might also want to replace recv_timeout by recv_deadline as a belt-and-braces solution?

Yes I guess in crossbeam terms the below could be done. And the IPC implementation would be doing `try_recv` in a loop while checking the timeout? 

The main point is that the timeout is not reset each time a message is received....

```rust
let deadline = after(timeout);
while true {
    select! {
        recv(r) -> msg => {},
        recv(after(timeout)) -> _ => break;,
    }
}
```

Also I think it's worth re-highlighting that the underlying problem is that WebXR is running a "sub message-loop" inside `perform_updates`, which itself is one part of the "main device loop". And while the sub-loop is running, the device loop is not, which means user input is not handled(At least in Hololens).

And the danger is that since the sub-loop is driven by messages coming from script, it could find a way to keep it running and prevent UI events from being handled.

**Comment by asajeffrey** [comment]
Well yes, really we should just have one main thread loop, but I'm not sure how to achieve that without a drastic rewrite. I'm going to open a "main thread" metabug over in servo/servo to try to capture the discussion, which is currently happening in lots of different issues.
